File: core/backend/app/policy/decisions.py
from __future__ import annotations
from dataclasses import dataclass, field
from datetime import datetime, UTC
from enum import Enum
from typing import Any, Optional


class Decision(str, Enum):
    ALLOW = "allow"
    DENY = "deny"
    REQUIRE_APPROVAL = "require_approval"


class RiskLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


@dataclass
class PolicyDecision:
    decision: Decision
    message: str
    risk_level: RiskLevel = RiskLevel.LOW
    reason_code: Optional[str] = None
    required_approver_role: Optional[str] = None
    policy_rule_id: Optional[str] = None
    policy_source: str = "builtin"
    policy_id: Optional[str] = None
    actor_type: Optional[str] = None
    actor_id: Optional[str] = None
    actor_ref: Optional[dict[str, Any]] = None
    space_id: Optional[str] = None
    action: Optional[str] = None
    resource_type: Optional[str] = None
    resource_id: Optional[str] = None
    audit_code: Optional[str] = None
    approval_capability: Optional[str] = None
    proposal_type: Optional[str] = None
    metadata_json: Optional[dict[str, Any]] = None
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    @property
    def allowed(self) -> bool:
        return self.decision == Decision.ALLOW

    @property
    def denied(self) -> bool:
        return self.decision == Decision.DENY

    @property
    def requires_approval(self) -> bool:
        return self.decision == Decision.REQUIRE_APPROVAL

File: control-plane/src/modules/policy/decisions.ts
/**
 * Policy decision value objects — TS port of `app/policy/decisions.py`
 * and the risk-rank helpers from `approval.py`.
 *
 * `PolicyDecision` mirrors the wire schema (`@agent-space/protocol` `policy.ts`)
 * field-for-field (snake_case) so a decision serializes directly to the
 * `PolicyDecisionSchema` contract.
 */

import type {
  PolicyDecisionValue,
  PolicyRiskLevel,
} from "@agent-space/protocol" with { "resolution-mode": "import" };

export type Decision = PolicyDecisionValue; // "allow" | "deny" | "require_approval"
export type RiskLevel = PolicyRiskLevel; // "low" | "medium" | "high" | "critical"

export interface PolicyDecision {
  decision: Decision;
  message: string;
  risk_level: RiskLevel;
  reason_code?: string | null;
  required_approver_role?: string | null;
  policy_rule_id?: string | null;
  policy_source: string;
  policy_id?: string | null;
  actor_type?: string | null;
  actor_id?: string | null;
  actor_ref?: Record<string, unknown> | null;
  space_id?: string | null;
  action?: string | null;
  resource_type?: string | null;
  resource_id?: string | null;
  audit_code?: string | null;
  approval_capability?: string | null;
  proposal_type?: string | null;
  metadata_json?: Record<string, unknown> | null;
  created_at?: string | null;
}

/**
 * Build a `PolicyDecision` with the same defaults as the Python dataclass
 * (`risk_level=low`, `policy_source=builtin`). Only `decision` and `message`
 * are required; everything else is optional.
 */
export function makeDecision(
  fields: Partial<PolicyDecision> & {
    decision: Decision;
    message: string;
  },
): PolicyDecision {
  return {
    risk_level: "low",
    policy_source: "builtin",
    ...fields,
  };
}

export function isAllowed(d: PolicyDecision): boolean {
  return d.decision === "allow";
}
export function isDenied(d: PolicyDecision): boolean {
  return d.decision === "deny";
}
export function requiresApproval(d: PolicyDecision): boolean {
  return d.decision === "require_approval";
}

export const RISK_RANK: Record<RiskLevel, number> = {
  low: 0,
  medium: 1,
  high: 2,
  critical: 3,
};

export const VALID_RISK_LEVELS: ReadonlySet<string> = new Set([
  "low",
  "medium",
  "high",
  "critical",
]);

/**
 * Mirror Python's `repr()` for strings (the `{x!r}` format used in policy
 * messages): single quotes by default, double quotes when the value contains a
 * single quote but no double quote, with backslash/quote escaping. Messages
 * must match Python byte-for-byte for cross-language decision parity.
 */
export function pyRepr(value: string): string {
  const hasSingle = value.includes("'");
  const hasDouble = value.includes('"');
  const quote = hasSingle && !hasDouble ? '"' : "'";
  const body = value
    .replace(/\\/g, "\\\\")
    .split(quote)
    .join("\\" + quote);
  return quote + body + quote;
}
